File: CS/OutlookExportImportSample/Form1.cs
using DevExpress.XtraScheduler;
using DevExpress.XtraScheduler.Outlook;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace OutlookExportImportSample {
    public partial class Form1 : DevExpress.XtraBars.Ribbon.RibbonForm {
        public Form1() {
            InitializeComponent();
            this.ribbonControl1.SelectedPage = this.ribbonPage1;
            this.schedulerStorage1.Appointments.ResourceSharing = true;
            this.schedulerControl1.GroupType = SchedulerGroupType.Resource;
        }

        private void Form1_Load(object sender, EventArgs e) {
            // TODO: This line of code loads data into the 'scheduleTestDataSet.Resources' table. You can move, or remove it, as needed.
            this.resourcesTableAdapter.Fill(this.scheduleTestDataSet.Resources);
            // TODO: This line of code loads data into the 'scheduleTestDataSet.Appointments' table. You can move, or remove it, as needed.
            this.appointmentsTableAdapter.Fill(this.scheduleTestDataSet.Appointments);
        }
        private void OnAppointmentChangedInsertedDeleted(object sender, PersistentObjectsEventArgs e) {
            appointmentsTableAdapter.Update(scheduleTestDataSet);
            scheduleTestDataSet.AcceptChanges();
        }

        private void barbtnOutlookExportSingle_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            #region #ExportSelectedAppointments
            AppointmentBaseCollection apts = schedulerControl1.SelectedAppointments;
            if (apts != null) {
                SchedulerStorage storage = new SchedulerStorage();
                foreach (Appointment apt in apts) {
                    Appointment aptCopy = apt.Copy();
                    storage.Appointments.Add(aptCopy);
                }
                storage.ExportToOutlook();
            }
            #endregion #ExportSelectedAppointments
        }

        private void barbtnExportUsingCriteria_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            lbcLog.Items.Clear();
            ExportUsingCriteria();
        }
        #region #ExportUsingCriteria
        private void ExportUsingCriteria() {
            OutlookExport exporter = schedulerControl1.Storage.CreateOutlookExporter() as OutlookExport;
            if (exporter != null) {
                exporter.AppointmentExporting += exporter_AppointmentExporting;
                exporter.AppointmentExported += exporter_AppointmentExported;
                exporter.OnException += exporter_OnException;
                exporter.CalendarFolderName = OutlookExchangeHelper.GetOutlookCalendarFolders().FirstOrDefault().FullPath;
                using (MemoryStream stream = new MemoryStream()) {
                    exporter.Export(stream);
                }
            }
        }
        void exporter_AppointmentExporting(object sender, AppointmentExportingEventArgs e) {
            AddToLog(String.Format("Exporting Subj:{0}, started at {1:F} ...", e.Appointment.Subject, e.Appointment.Start));
            if (e.Appointment.IsRecurring) {
                e.Cancel = true;
                AddToLog("Cancelled because of its type (recurring).");
            }
        }
        void exporter_AppointmentExported(object sender, AppointmentExportedEventArgs e) {
            AddToLog(String.Format("Successfully exported Subj:{0}, started at {1:F}!", e.Appointment.Subject, e.Appointment.Start)); 
        }
        void exporter_OnException(object sender, ExchangeExceptionEventArgs e) {
            string errText = e.OriginalException.Message;
            AddToLog(errText);
            OutlookExport exporter = (OutlookExport)sender;
            exporter.Terminate();
            e.Handled = true;
            //throw e.OriginalException;
        }
        #endregion #ExportUsingCriteria

        private void AddToLog(string logText) {
            lbcLog.Items.Add(logText);
        }

        private void barbtnOutlookImport_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            lbcLog.Items.Clear();
            ImportUsingCriteria();
        }

        #region #ImportUsingCriteria
        private void ImportUsingCriteria() {
            OutlookImport importer = schedulerControl1.Storage.CreateOutlookImporter() as OutlookImport;
            if (importer != null) {
                importer.AppointmentImporting += importer_AppointmentImporting;
                importer.AppointmentImported += importer_AppointmentImported;
                importer.OnException += importer_OnException;
                importer.CalendarFolderName = OutlookExchangeHelper.GetOutlookCalendarFolders().FirstOrDefault().FullPath;
                using (MemoryStream stream = new MemoryStream()) {
                    importer.Import(stream);
                }
            }
        }
        void importer_AppointmentImporting(object sender, AppointmentImportingEventArgs e) {
            OutlookAppointmentImportingEventArgs args = e as OutlookAppointmentImportingEventArgs;
            AddToLog(String.Format("Importing Subj:{0}, started at {1:F} ...", args.OutlookAppointment.Subject, args.OutlookAppointment.Start));
            if (args.OutlookAppointment.BusyStatus == DevExpress.XtraScheduler.Outlook.Interop.OlBusyStatus.olWorkingElsewhere) {
                e.Cancel = true;
                AddToLog("Cancelled because of its busy type (working elsewhere).");
            }
        }
        void importer_AppointmentImported(object sender, AppointmentImportedEventArgs e) {
            OutlookAppointmentImportedEventArgs args = e as OutlookAppointmentImportedEventArgs;
            AddToLog(String.Format("Successfully imported Subj:{0}, started at {1:F}!", args.OutlookAppointment.Subject, args.OutlookAppointment.Start)); 
        }

        void importer_OnException(object sender, ExchangeExceptionEventArgs e) {
            string errText = e.OriginalException.Message;
            AddToLog(errText);
            OutlookImport importer = (OutlookImport)sender;
            importer.Terminate();
            e.Handled = true;
            //throw e.OriginalException;
        }
        #endregion #ImportUsingCriteria
    }
}

File: VB/OutlookExportImportSample/Form1.vb
Imports DevExpress.XtraScheduler
Imports DevExpress.XtraScheduler.Outlook
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Namespace OutlookExportImportSample
    Partial Public Class Form1
        Inherits DevExpress.XtraBars.Ribbon.RibbonForm

        Public Sub New()
            InitializeComponent()
            Me.ribbonControl1.SelectedPage = Me.ribbonPage1
            Me.schedulerStorage1.Appointments.ResourceSharing = True
            Me.schedulerControl1.GroupType = SchedulerGroupType.Resource
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            ' TODO: This line of code loads data into the 'scheduleTestDataSet.Resources' table. You can move, or remove it, as needed.
            Me.resourcesTableAdapter.Fill(Me.scheduleTestDataSet.Resources)
            ' TODO: This line of code loads data into the 'scheduleTestDataSet.Appointments' table. You can move, or remove it, as needed.
            Me.appointmentsTableAdapter.Fill(Me.scheduleTestDataSet.Appointments)
        End Sub
        Private Sub OnAppointmentChangedInsertedDeleted(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs) Handles schedulerStorage1.AppointmentsInserted, schedulerStorage1.AppointmentsChanged, schedulerStorage1.AppointmentsDeleted
            appointmentsTableAdapter.Update(scheduleTestDataSet)
            scheduleTestDataSet.AcceptChanges()
        End Sub

        Private Sub barbtnOutlookExportSingle_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barbtnOutlookExportSelected.ItemClick
'            #Region "#ExportSelectedAppointments"
            Dim apts As AppointmentBaseCollection = schedulerControl1.SelectedAppointments
            If apts IsNot Nothing Then
                Dim storage As New SchedulerStorage()
                For Each apt As Appointment In apts
                    Dim aptCopy As Appointment = apt.Copy()
                    storage.Appointments.Add(aptCopy)
                Next apt
                storage.ExportToOutlook()
            End If
'            #End Region ' #ExportSelectedAppointments
        End Sub

        Private Sub barbtnExportUsingCriteria_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barbtnExportUsingCriteria.ItemClick
            lbcLog.Items.Clear()
            ExportUsingCriteria()
        End Sub
        #Region "#ExportUsingCriteria"
        Private Sub ExportUsingCriteria()
            Dim exporter As OutlookExport = TryCast(schedulerControl1.Storage.CreateOutlookExporter(), OutlookExport)
            If exporter IsNot Nothing Then
                AddHandler exporter.AppointmentExporting, AddressOf exporter_AppointmentExporting
                AddHandler exporter.AppointmentExported, AddressOf exporter_AppointmentExported
                AddHandler exporter.OnException, AddressOf exporter_OnException
                exporter.CalendarFolderName = OutlookExchangeHelper.GetOutlookCalendarFolders().FirstOrDefault().FullPath
                Using stream As New MemoryStream()
                    exporter.Export(stream)
                End Using
            End If
        End Sub
        Private Sub exporter_AppointmentExporting(ByVal sender As Object, ByVal e As AppointmentExportingEventArgs)
            AddToLog(String.Format("Exporting Subj:{0}, started at {1:F} ...", e.Appointment.Subject, e.Appointment.Start))
            If e.Appointment.IsRecurring Then
                e.Cancel = True
                AddToLog("Cancelled because of its type (recurring).")
            End If
        End Sub
        Private Sub exporter_AppointmentExported(ByVal sender As Object, ByVal e As AppointmentExportedEventArgs)
            AddToLog(String.Format("Successfully exported Subj:{0}, started at {1:F}!", e.Appointment.Subject, e.Appointment.Start))
        End Sub
        Private Sub exporter_OnException(ByVal sender As Object, ByVal e As ExchangeExceptionEventArgs)
            Dim errText As String = e.OriginalException.Message
            AddToLog(errText)
            Dim exporter As OutlookExport = DirectCast(sender, OutlookExport)
            exporter.Terminate()
            e.Handled = True
            'throw e.OriginalException;
        End Sub
        #End Region ' #ExportUsingCriteria

        Private Sub AddToLog(ByVal logText As String)
            lbcLog.Items.Add(logText)
        End Sub

        Private Sub barbtnOutlookImport_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barbtnOutlookImport.ItemClick
            lbcLog.Items.Clear()
            ImportUsingCriteria()
        End Sub

        #Region "#ImportUsingCriteria"
        Private Sub ImportUsingCriteria()
            Dim importer As OutlookImport = TryCast(schedulerControl1.Storage.CreateOutlookImporter(), OutlookImport)
            If importer IsNot Nothing Then
                AddHandler importer.AppointmentImporting, AddressOf importer_AppointmentImporting
                AddHandler importer.AppointmentImported, AddressOf importer_AppointmentImported
                AddHandler importer.OnException, AddressOf importer_OnException
                importer.CalendarFolderName = OutlookExchangeHelper.GetOutlookCalendarFolders().FirstOrDefault().FullPath
                Using stream As New MemoryStream()
                    importer.Import(stream)
                End Using
            End If
        End Sub
        Private Sub importer_AppointmentImporting(ByVal sender As Object, ByVal e As AppointmentImportingEventArgs)
            Dim args As OutlookAppointmentImportingEventArgs = TryCast(e, OutlookAppointmentImportingEventArgs)
            AddToLog(String.Format("Importing Subj:{0}, started at {1:F} ...", args.OutlookAppointment.Subject, args.OutlookAppointment.Start))
            If args.OutlookAppointment.BusyStatus = DevExpress.XtraScheduler.Outlook.Interop.OlBusyStatus.olWorkingElsewhere Then
                e.Cancel = True
                AddToLog("Cancelled because of its busy type (working elsewhere).")
            End If
        End Sub
        Private Sub importer_AppointmentImported(ByVal sender As Object, ByVal e As AppointmentImportedEventArgs)
            Dim args As OutlookAppointmentImportedEventArgs = TryCast(e, OutlookAppointmentImportedEventArgs)
            AddToLog(String.Format("Successfully imported Subj:{0}, started at {1:F}!", args.OutlookAppointment.Subject, args.OutlookAppointment.Start))
        End Sub

        Private Sub importer_OnException(ByVal sender As Object, ByVal e As ExchangeExceptionEventArgs)
            Dim errText As String = e.OriginalException.Message
            AddToLog(errText)
            Dim importer As OutlookImport = DirectCast(sender, OutlookImport)
            importer.Terminate()
            e.Handled = True
            'throw e.OriginalException;
        End Sub
        #End Region ' #ImportUsingCriteria
    End Class
End Namespace
